cadc-gms/src/main/java/org/opencadc/auth/StandardIdentityManager.java

```diff
@@ -0,0 +1,272 @@
+import ca.nrc.cadc.util.StringUtil;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.security.Principal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import javax.security.auth.Subject;
+import javax.security.auth.x500.X500Principal;
+import org.apache.log4j.Logger;
+import org.json.JSONObject;
+
+/**
+ * Prototype IdentityManager for a standards-based system.
+ * 
+ * @author pdowler
+ */
+public class StandardIdentityManager implements IdentityManager {
+    private static final Logger log = Logger.getLogger(StandardIdentityManager.class);
+
+    private final URI oidcIssuer;
+    
+    // TODO: need these to contruct an AuthorizationToken
+    private List<String> oidcDomains = new ArrayList<>();
+    private URI oidcScope;
+    
+    public StandardIdentityManager() {
+        LocalAuthority loc = new LocalAuthority();
+        String key = Standards.SECURITY_METHOD_OPENID.toASCIIString();
+        this.oidcIssuer = loc.getServiceURI(key);
+        try {
+            URL u = oidcIssuer.toURL();
+            oidcDomains.add(u.getHost());
+        } catch (MalformedURLException ex) {
+            throw new InvalidConfigException("found " + key + " = " + oidcIssuer + " - expected valid URL", ex);
+        }
+    }
+
+    @Override
+    public Subject validate(Subject subject) throws NotAuthenticatedException {
+        validateOidcAccessToken(subject);
+        return subject;
+    }
+
+    @Override
+    public Subject augment(Subject subject) {
+        // TODO: if X500Principal && servops && UMS we could augment CADC-style
+        // oidc tokens: validate gets HttpPrincipal and NumericPrincial
+        // cadc signed cookies/tokens: validate gets all identities
+        return subject;
+    }
+
+    @Override
+    public Subject toSubject(Object owner) {
+        Subject ret = new Subject();
+        if (owner != null) {
+            UUID uuid = null;
+            if (owner instanceof UUID) {
+                uuid = (UUID) owner;
+            } else if (owner instanceof String) {
+                String sub = (String) owner;
+                uuid = UUID.fromString(sub);
+            } else {
+                throw new RuntimeException("unexpected owner type: " + owner.getClass().getName() + " value: " + owner);
+            }
+            NumericPrincipal p = new NumericPrincipal(uuid);
+            
+            // effectively augment by using the current subject as a "cache" of known identities
+            Subject s = AuthenticationUtil.getCurrentSubject();
+            if (s != null) {
+                for (Principal cp : s.getPrincipals()) {
+                    if (AuthenticationUtil.equals(p, cp)) {
+                        log.debug("[cache hit] caller Subject matches " + p + ": " + s);
+                        ret.getPrincipals().addAll(s.getPrincipals());
```

Comment:
I guess all the principals are added because if we match on any (check on line 155), then we can trust that a previous augment subject was done correctly.

Reply:
yup; unlike ACIdentityManager I didn't just return the cached subject (because it has other junk in it) so copied the principals instead.